File: 2368.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {

    private List<Integer>[] graph;
    private boolean[][] visited_plus_restricted;
    private int count;
    public int reachableNodes(int n, int[][] edges, int[] restricted) {
        graph = adjencyList(edges, n);
        visited_plus_restricted = new boolean[n][2]; // {{visited, restricted}, ...edges.}
        for (int r : restricted) { // restricted nodes 
            visited_plus_restricted[r][1] = true;
        }
        bfs(0);

        return count;

    }
    public void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited_plus_restricted[start][0] = true;
        ++count;

        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int u : graph[node]) {
                if (!visited_plus_restricted[u][0] && !visited_plus_restricted[u][1]) {
                    visited_plus_restricted[u][0] = true;
                    ++count;
                    queue.add(u);
                }
            }
        }
    }
    static List<Integer>[] adjencyList(int[][] edges, int n) {
        List<Integer>[] adj_list = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj_list[i] = new ArrayList<>();
        }
        for (int[] edge : edges) {
            adj_list[edge[0]].add(edge[1]);
            adj_list[edge[1]].add(edge[0]);
        }
        return adj_list;

    }

}

File: 2368.py
'''
Objective: find the maximum number of nodes in the tree that can be visited, starting from node 0, 
           without entering the restricted nodes.
Idea of solution: use BFS starting at node 0, while skipping all restricted nodes. count all the reachable nodes 
                  that are not on the restricted list and can be reched without restrictions.
Complexity: O(n)
'''

from collections import deque

class Solution:
    def reachableNodes(self, n, edges, restricted):
        restricted = set(restricted) #for fast search
        
        # initialize the graph as an adjacency list
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        visited = [False] * n # an array for marking visited nodes (initially all False)

        # queue for BFS, starting from node 0. mark 0 as visited (visited[0] = True)
        # count is answer - count of reachable nodes

        queue = deque()
        queue.append(0)
        visited[0] = True
        count = 0
        
        # BFS algorithm
        
        while queue:
            
            # while the queue is not empty -> extract the cur node from the beginning of the queue.
            # increasing the count counter, since cur is the new reachable node.

            cur = queue.popleft()
            count += 1
            
            # processing the neighbors of the current node

            for neighbor in adj[cur]:
                
                # if the neighbor is not visited and not restricted -> next cur node for BFS -> mark it as visited 
                # and add it to the queue => visit all accessible nodes, skip the forbidden ones

                if not visited[neighbor] and neighbor not in restricted:
                    visited[neighbor] = True
                    queue.append(neighbor)
        
        return count